### PR TITLE
Handle do not match columns and values

### LINKED ISSUE #189: Throw on too many values
I've just noticed that this doesn't error, that's a problem:
```sql
INSERT INTO TestUnique (name) VALUES ('test1'), ('test2', 3)
```
Instead it acts like:
```sql
INSERT INTO TestUnique (name) VALUES ('test1'), ('test2')
```

### PR DESCRIPTION
**Pull request by ever0de**
fix #189

### CODE REVIEW

File: src/data/row.rs [hunk 1]
```diff
@@ -41,34 +41,40 @@ impl Row {
     }
 
     pub fn new(column_defs: &[ColumnDef], columns: &[String], values: &[Expr]) -> Result<Self> {
+        if !columns.is_empty() && values.len() != columns.len() {
+            return Err(RowError::DoNotMatchColumnsAndValues.into());
+        }
         if values.len() > column_defs.len() {
```
Comment (panarch) [inline src/data/row.rs:47]: It would be good to use `else if`

File: src/data/row.rs [hunk 2]
```diff
@@ -41,34 +41,40 @@ impl Row {
...
             return Err(RowError::TooManyValues.into());
         }
 
+        let columns: Box<dyn Iterator<Item = &String>> = if columns.is_empty() {
+            Box::new(column_defs.iter().map(|ColumnDef { name, .. }| name))
+        } else {
+            Box::new(columns.iter())
+        };
```
Comment (panarch) [inline src/data/row.rs:55]: `iter_enum::Iterator` would be quite a good choice to take it here.

```rust
#[derive(Iterator)]                                                    
enum Columns<I1, I2> {                                                 
    All(I1),                                                           
    Specified(I2),                                                     
}                                                                      
                                                                       
let columns = if columns.is_empty() {                                  
    Columns::All(column_defs.iter().map(|ColumnDef { name, .. }| name))
} else {                                                               
    Columns::Specified(columns.iter())                                 
};                                                                     
```

File: src/data/row.rs [hunk 3]
```diff
@@ -15,8 +15,8 @@ pub enum RowError {
     #[error("lack of required column: {0}")]
     LackOfRequiredColumn(String),
 
-    #[error("literals does not fit to columns")]
-    LackOfRequiredValue(String),
+    #[error("do not match columns and values")]
+    DoNotMatchColumnsAndValues,
```
Comment (panarch) [inline src/data/row.rs:19]: Error name or message is for describing what's wrong happened. so, imperative sentence might not be a good choice.
How about simply `ColumnAndValuesNotMatched`?

**Review by panarch — CHANGES_REQUESTED** [review]
Looks really nice, thanks!
Just a few tiny comments.

File: src/data/row.rs (hunk 2 above)
Comment (ever0de) [reply src/data/row.rs]: A very nice change! thank you.

File: src/data/row.rs (hunk 3 above)
Comment (ever0de) [reply src/data/row.rs]: That's a good comment. changed it

File: src/data/row.rs (hunk 1 above)
Comment (ever0de) [reply src/data/row.rs:54]: It seems like a good change. I'll fix it

File: src/data/row.rs [hunk 4]
```diff
@@ -25,6 +25,13 @@ pub enum RowError {
     ConflictOnEmptyRow,
 }
 
+#[derive(iter_enum::Iterator)]
+
+enum Columns<I1, I2> {
```
Comment (panarch) [inline src/data/row.rs:30]: Let's just remove this blank line, then all looks good!

Comment (ever0de) [reply src/data/row.rs:30]: Uh.. why did this go in? will do fixup

**Review by panarch — APPROVED** [review]
Thanks! merging 👍

**Pull request merged**